marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.hub/lib/hub-utils-lib.xqy

```diff
@@ -86,21 +86,44 @@ declare function hul:invalidate-field-cache($name)
   xdmp:set-server-field($name || "-refresh-date", ())
 };
 
+(:  faster cache function. Won't take a closure, so caller has to 
+ :  call hul:set-field-cache($name, $func(), $duration) to re-initialize if empty
+ :  Likely faster because caller does not have to build a closure. 
+ :)
+declare function hul:from-field-cache-or-empty($name, $duration as xs:dayTimeDuration?) {
+  let $existing := xdmp:get-server-field($name)
+  let $still-fresh := empty($duration) or (fn:current-dateTime() < xdmp:get-server-field($name || "-refresh-date"))
+  return
+    if(exists($existing) and $still-fresh) 
+    then $existing
+    else 
+      let$_ := hul:invalidate-field-cache($name)
+      return ()
+}; 
+
+(:~
+ : Get a cahced value, but if missing or expired invoke $func to get the value and cache it. 
+ :)
 declare function hul:from-field-cache($name, $func, $duration as xs:dayTimeDuration?)
 {
   let $existing := xdmp:get-server-field($name)
   let $refresh-date := xdmp:get-server-field($name || "-refresh-date")
+  (: update the cache and duration if any of these hold:
+       cache is empty
+       duration has expired 
+       there is now a duration supplied and an eternal/noduration entry is already cached
+       :)
   return
-    if (fn:exists($existing)) then
-      if (fn:exists($refresh-date) and ($refresh-date < fn:current-dateTime())) then
-        hul:set-field-cache($name, $func(), $duration)
-      else
+    if ( (exists($refresh-date) and ($refresh-date < fn:current-dateTime() + xs:dayTimeDuration("PT0."||xdmp:random(9)||"S")))
+         or empty($existing)
+         or exists($duration) and empty($refresh-date) 
+        )
+      then 
+        ( xdmp:log("field cache miss xx : "||$name), 
```

Comment:
generally we've been avoiding xdmp:logs(). Do we need to log this out? or is this leftover debugging code?